{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "548ad0a3-cbe2-42d4-989c-3a101f694615",
   "metadata": {},
   "source": [
    "# Reads assignment example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e0e9b88e-12d3-415d-8721-7a4bc3455762",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from starmap.sequencing import *\n",
    "from natsort import natsorted\n",
    "from scipy.io import loadmat, savemat\n",
    "from skimage.filters import threshold_otsu\n",
    "from skimage.color import label2rgb\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c10bf5b-2553-4f33-ad46-1d1ae2eb0b4e",
   "metadata": {},
   "source": [
    "### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d4a054bf-1aec-477d-9a41-5ee3cfe235c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import wraps\n",
    "from time import time\n",
    "\n",
    "# Timer\n",
    "def timer(func):\n",
    "    @wraps(func)\n",
    "    def _time_it(*args, **kwargs):\n",
    "        start = int(round(time() * 1000))\n",
    "        try:\n",
    "            return func(*args, **kwargs)\n",
    "        finally:\n",
    "            end_ = int(round(time() * 1000)) - start\n",
    "            end_ = round(end_ / 1000, 4)\n",
    "            print(f\"Total execution time: {end_ if end_ > 0 else 0} s\")\n",
    "    return _time_it\n",
    "\n",
    "\n",
    "@timer\n",
    "# Load reads and their positions from mat file\n",
    "def load_reads(fpath, reads_file):\n",
    "    S = loadmat(os.path.join(fpath, reads_file))\n",
    "    bases = [str(i[0][0]) for i in S[\"merged_reads\"]]\n",
    "    points = S[\"merged_points\"]\n",
    "    temp = np.zeros(points.shape)\n",
    "    temp[:, 0] = np.round(points[:, 1]-1)\n",
    "    temp[:, 1] = np.round(points[:, 0]-1)\n",
    "    temp[:, 2] = np.round(points[:, 2]-1)\n",
    "    \n",
    "    print(f\"Number of reads: {len(bases)}\")\n",
    "    \n",
    "    return bases, temp\n",
    "\n",
    "\n",
    "# Load gene table from genes.csv\n",
    "def load_genes(fpath):\n",
    "    genes2seq = {}\n",
    "    seq2genes = {}\n",
    "    with open(os.path.join(fpath, \"genes.csv\"), encoding='utf-8-sig') as f:\n",
    "        for l in f:\n",
    "            fields = l.rstrip().split(\",\")\n",
    "            genes2seq[fields[0]] = \"\".join([str(s+1) for s in encode_SOLID(fields[1][::-1])])\n",
    "            seq2genes[genes2seq[fields[0]]] = fields[0]\n",
    "    return genes2seq, seq2genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11b496c1-d015-4bd7-9a28-7dc3d7238199",
   "metadata": {},
   "source": [
    "## Input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "ed4aaf5a-8a3b-47e0-84e0-a47c1d464856",
   "metadata": {},
   "outputs": [],
   "source": [
    "# IO path \n",
    "# base_path = 'Z:/Data/Analyzed/2021-08-25-Hu-HelaSTARmap'\n",
    "base_path = 'Z:/Data/Analyzed/2021-07-09-Hu-HelaRIBOmap'\n",
    "\n",
    "out_path = os.path.join(base_path, 'output')\n",
    "if not os.path.exists(out_path):\n",
    "    os.mkdir(out_path)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "790ca72c-f0d3-4c57-b809-4606bc7b8cdb",
   "metadata": {},
   "source": [
    "## Generate reads pattern plot "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d82ac3c4-ab4b-4fbb-b2db-a747714afcf7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>AARS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>ASXL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>COX7B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>CTGF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>DLX4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>976</th>\n",
       "      <td>0</td>\n",
       "      <td>ZRANB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>977</th>\n",
       "      <td>3</td>\n",
       "      <td>ZSCAN12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>978</th>\n",
       "      <td>1</td>\n",
       "      <td>ZSCAN29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>979</th>\n",
       "      <td>3</td>\n",
       "      <td>ZXDC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>980</th>\n",
       "      <td>0</td>\n",
       "      <td>ZYG11A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>981 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     cluster     gene\n",
       "0          0     AARS\n",
       "1          0    ASXL1\n",
       "2          0    COX7B\n",
       "3          1     CTGF\n",
       "4          0     DLX4\n",
       "..       ...      ...\n",
       "976        0   ZRANB1\n",
       "977        3  ZSCAN12\n",
       "978        1  ZSCAN29\n",
       "979        3     ZXDC\n",
       "980        0   ZYG11A\n",
       "\n",
       "[981 rows x 2 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load gene cluters\n",
    "gene_cluster_df = pd.read_csv('../other-datasets/gene_cluster_label.csv')\n",
    "gene_cluster_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ee146648-8ede-43e3-857d-bd108890baac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current sample: Z:/Data/Analyzed/2021-07-09-Hu-HelaRIBOmap\n",
      "Number of reads: 5189148\n",
      "Total execution time: 36.272 s\n",
      "Current cluster: 0\n",
      "Current cluster: 1\n",
      "Current cluster: 2\n",
      "Current cluster: 3\n"
     ]
    }
   ],
   "source": [
    "color_dict = {0:'green', 1:'yellow', 2:'white', 3:'magenta'}\n",
    "\n",
    "print(f\"Current sample: {base_path}\")\n",
    "\n",
    "# Load genes\n",
    "genes2seqs, seqs2genes = load_genes(base_path)\n",
    "\n",
    "# Load reads \n",
    "bases, points = load_reads(base_path, \"2021-10-03-merged_goodPoints_max3d\")\n",
    "bases = list(map(lambda x: seqs2genes[x], bases))\n",
    "bases = np.array(bases)\n",
    "points = np.array(points)\n",
    "\n",
    "fig_out_path = os.path.join(os.path.join(base_path, 'figures'))\n",
    "if not os.path.exists(fig_out_path):\n",
    "    os.mkdir(fig_out_path)\n",
    "\n",
    "# Load raw image\n",
    "# current_img = load_label_image(os.path.join(base_path, 'images'), fname='overlay_max.tif')\n",
    "current_img = load_label_image(os.path.join(base_path, 'segmentation'), fname='cell_bnd.tif')\n",
    "\n",
    "# Load segmentation\n",
    "current_seg = load_label_image(os.path.join(base_path, 'segmentation'), fname='cell.tif')\n",
    "\n",
    "points = points.astype(int)\n",
    "reads_assignment = current_seg[points[:, 2], points[:, 0], points[:, 1]]\n",
    "\n",
    "reads_info = pd.DataFrame({'x':points[:, 0], 'y':points[:, 1], 'z':points[:, 2], 'cell_label':reads_assignment})\n",
    "reads_info = reads_info.astype(np.int32)\n",
    "reads_info['orig_index'] = reads_info['cell_label'] - 1\n",
    "reads_info['gene'] = bases\n",
    "\n",
    "for current_cluster in sorted(gene_cluster_df['cluster'].unique()):\n",
    "    print(f\"Current cluster: {current_cluster}\")\n",
    "    current_gene_list = gene_cluster_df.loc[gene_cluster_df['cluster'] == current_cluster, 'gene'].to_list()\n",
    "    current_reads_df = reads_info.loc[reads_info['gene'].isin(current_gene_list), :]\n",
    "\n",
    "    # remove unassigned reads\n",
    "    current_reads_df = current_reads_df.loc[current_reads_df['cell_label'] != 0, :]\n",
    "\n",
    "    plt.figure(figsize=(current_img.shape[0]/1000, current_img.shape[1]/1000), dpi=1000)\n",
    "    plt.imshow(current_img, cmap='gray')\n",
    "    plt.plot(current_reads_df['y'], current_reads_df['x'], '.', color=color_dict[current_cluster], markersize=.3, markeredgewidth=0.0)\n",
    "    plt.axis('off')\n",
    "    plt.tight_layout(pad=0)\n",
    "    # plt.show()\n",
    "    current_fig_path = f\"{fig_out_path}/cluster_{current_cluster}.tiff\"\n",
    "    plt.savefig(current_fig_path, dpi=1000, pil_kwargs={\"compression\": \"tiff_lzw\"}, bbox_inches='tight', pad_inches=0)\n",
    "    plt.clf()\n",
    "    plt.close()\n",
    "\n",
    "plt.figure(figsize=(current_img.shape[0]/1000, current_img.shape[1]/1000), dpi=1000)\n",
    "plt.imshow(current_img, cmap='gray')\n",
    "for current_cluster in sorted(gene_cluster_df['cluster'].unique()):\n",
    "    current_gene_list = gene_cluster_df.loc[gene_cluster_df['cluster'] == current_cluster, 'gene'].to_list()\n",
    "    current_reads_df = reads_info.loc[reads_info['gene'].isin(current_gene_list), :]\n",
    "\n",
    "    # remove unassigned reads\n",
    "    current_reads_df = current_reads_df.loc[current_reads_df['cell_label'] != 0, :]\n",
    "\n",
    "    plt.plot(current_reads_df['y'], current_reads_df['x'], '.', color=color_dict[current_cluster], markersize=.3, markeredgewidth=0.0)\n",
    "plt.axis('off')\n",
    "plt.tight_layout(pad=0)\n",
    "# plt.show()\n",
    "current_fig_path = f\"{fig_out_path}/cluster_all.tiff\"\n",
    "plt.savefig(current_fig_path, dpi=1000, pil_kwargs={\"compression\": \"tiff_lzw\"}, bbox_inches='tight', pad_inches=0)\n",
    "plt.clf()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "222f6385-39c2-4f36-afb2-f5216a0cfda5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
